sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -922,6 +923,29 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
   /**
    * Returns the partition names from hive metastore for a given table in a database.
    */
+  override def listPartitionNames(
+      db: String,
+      table: String,
+      partialSpec: Option[TablePartitionSpec] = None): Seq[String] = withClient {
+    val actualPartColNames = getTable(db, table).partitionColumnNames
+    val clientPartitionNames =
+      client.getPartitionNames(db, table, partialSpec.map(lowerCasePartitionSpec))
+
+    if (actualPartColNames.exists(partColName => partColName != partColName.toLowerCase)) {
+      clientPartitionNames.map { partName =>
+        val partSpec = PartitioningUtils.parsePathFragmentAsSeq(partName)
+        partSpec.map { case (partName, partValue) =>
+          actualPartColNames.find(_.equalsIgnoreCase(partName)).get + "=" + partValue
+        }.mkString("/")
+      }
+    } else {
+      clientPartitionNames
```

Comment:
I ran `show partitions` on a table with about 99k partitions using a build with and without this short circuit. I didn't keep count, but I must have run the command at least 10 times in each session. The best time I got for listing with the short circuit was 1.496 seconds. The best time I got for listing without the short circuit was 1.671.

I can't say I feel strongly one way or the other. I guess my vote is to keep the short circuit.